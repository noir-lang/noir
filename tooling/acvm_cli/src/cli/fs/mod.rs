pub(super) mod inputs;
pub(super) mod witness;
