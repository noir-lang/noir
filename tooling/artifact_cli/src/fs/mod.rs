pub mod artifact;
pub mod inputs;
pub mod witness;
