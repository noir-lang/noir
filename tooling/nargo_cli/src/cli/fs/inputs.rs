use noirc_abi::{
    input_parser::{Format, InputValue},
    Abi, InputMap, MAIN_RETURN_NAME,
};
use std::{collections::BTreeMap, path::Path};

use crate::errors::FilesystemError;

/// Returns the circuit's parameters and its return value, if one exists.
/// # Examples
///
/// ```ignore
/// let (input_map, return_value): (InputMap, Option<InputValue>) =
///   read_inputs_from_file(path, "Verifier", Format::Toml, &abi)?;
/// ```
pub(crate) fn read_inputs_from_file<P: AsRef<Path>>(
    path: P,
    file_name: &str,
    format: Format,
    abi: &Abi,
) -> Result<(InputMap, Option<InputValue>), FilesystemError> {
    if abi.is_empty() {
        return Ok((BTreeMap::new(), None));
    }

    let file_path = path.as_ref().join(file_name).with_extension(format.ext());
    if !file_path.exists() {
        return Err(FilesystemError::MissingTomlFile(file_name.to_owned(), file_path));
    }

    let input_string = std::fs::read_to_string(file_path).unwrap();
    let mut input_map = format.parse(&input_string, abi)?;
    let return_value = input_map.remove(MAIN_RETURN_NAME);

    Ok((input_map, return_value))
}
