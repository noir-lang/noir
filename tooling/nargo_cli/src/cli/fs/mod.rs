use std::{
    fs::File,
    io::Write,
    path::{Path, PathBuf},
};

pub(super) mod inputs;
pub(super) mod program;
pub(super) mod witness;

pub(super) fn create_named_dir(named_dir: &Path, name: &str) -> PathBuf {
    std::fs::create_dir_all(named_dir)
        .unwrap_or_else(|_| panic!("could not create the `{name}` directory"));

    PathBuf::from(named_dir)
}

pub(super) fn write_to_file(bytes: &[u8], path: &Path) -> String {
    let display = path.display();

    let mut file = match File::create(path) {
        Err(why) => panic!("couldn't create {display}: {why}"),
        Ok(file) => file,
    };

    match file.write_all(bytes) {
        Err(why) => panic!("couldn't write to {display}: {why}"),
        Ok(_) => display.to_string(),
    }
}
