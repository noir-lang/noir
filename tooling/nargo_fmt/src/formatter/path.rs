use noirc_frontend::{
    ast::{Path, PathKind, UnresolvedType},
    token::{Keyword, Token},
};

use super::Formatter;

impl<'a> Formatter<'a> {
    pub(super) fn format_path(&mut self, path: Path) {
        self.skip_comments_and_whitespace();

        match path.kind {
            PathKind::Plain => (),
            PathKind::Crate => {
                self.write_keyword(Keyword::Crate);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Dep => {
                self.write_keyword(Keyword::Dep);
                self.write_token(Token::DoubleColon);
            }
            PathKind::Super => {
                self.write_keyword(Keyword::Super);
                self.write_token(Token::DoubleColon);
            }
        }

        for (index, segment) in path.segments.into_iter().enumerate() {
            if index > 0 {
                self.write_token(Token::DoubleColon);
            }
            self.write_identifier(segment.ident);

            if let Some(generics) = segment.generics {
                self.format_turbofish(generics);
            }
        }
    }

    pub(super) fn format_turbofish(&mut self, generics: Vec<UnresolvedType>) {
        self.write_token(Token::DoubleColon);
        self.write_token(Token::Less);
        for (index, typ) in generics.into_iter().enumerate() {
            if index > 0 {
                self.write_comma();
                self.write_space();
            }
            self.format_type(typ);
        }

        // Skip trailing comma, if any
        self.skip_comments_and_whitespace();
        if self.is_at(Token::Comma) {
            self.bump();
        }

        self.write_token(Token::Greater);
    }
}
