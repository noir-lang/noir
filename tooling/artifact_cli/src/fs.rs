use std::path::Path;

use color_eyre::eyre::{self, Context};
use noirc_artifacts::{contract::ContractArtifact, program::ProgramArtifact};

/// A parsed JSON build artifact.
#[derive(Debug, Clone)]
pub enum Artifact {
    Program(ProgramArtifact),
    Contract(ContractArtifact),
}

impl Artifact {
    /// Try to parse an artifact as a binary program or a contract
    pub fn read_from_file(path: &Path) -> eyre::Result<Artifact> {
        let json = std::fs::read(path)
            .with_context(|| format!("failed to read artifact from '{}'", path.display()))?;

        let as_program = || serde_json::from_slice::<ProgramArtifact>(&json).map(Artifact::Program);
        let as_contract =
            || serde_json::from_slice::<ContractArtifact>(&json).map(Artifact::Contract);

        Ok(as_program().or_else(|e| as_contract().map_err(|_| e))?)
    }
}
