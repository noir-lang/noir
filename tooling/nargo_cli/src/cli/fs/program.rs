use std::path::{Path, PathBuf};

use nargo::artifacts::{contract::ContractArtifact, program::ProgramArtifact};
use noirc_frontend::graph::CrateName;

use crate::errors::FilesystemError;

use super::{create_named_dir, write_to_file};

pub(crate) fn save_program_to_file<P: AsRef<Path>>(
    program_artifact: &ProgramArtifact,
    crate_name: &CrateName,
    circuit_dir: P,
) -> PathBuf {
    let circuit_name: String = crate_name.into();
    save_build_artifact_to_file(program_artifact, &circuit_name, circuit_dir)
}

pub(crate) fn save_contract_to_file<P: AsRef<Path>>(
    compiled_contract: &ContractArtifact,
    circuit_name: &str,
    circuit_dir: P,
) -> PathBuf {
    save_build_artifact_to_file(compiled_contract, circuit_name, circuit_dir)
}

fn save_build_artifact_to_file<P: AsRef<Path>, T: ?Sized + serde::Serialize>(
    build_artifact: &T,
    artifact_name: &str,
    circuit_dir: P,
) -> PathBuf {
    create_named_dir(circuit_dir.as_ref(), "target");
    let circuit_path = circuit_dir.as_ref().join(artifact_name).with_extension("json");

    write_to_file(&serde_json::to_vec(build_artifact).unwrap(), &circuit_path);

    circuit_path
}

pub(crate) fn read_program_from_file<P: AsRef<Path>>(
    circuit_path: P,
) -> Result<ProgramArtifact, FilesystemError> {
    let file_path = circuit_path.as_ref().with_extension("json");

    let input_string =
        std::fs::read(&file_path).map_err(|_| FilesystemError::PathNotValid(file_path))?;
    let program = serde_json::from_slice(&input_string)
        .map_err(|err| FilesystemError::ProgramSerializationError(err.to_string()))?;

    Ok(program)
}
