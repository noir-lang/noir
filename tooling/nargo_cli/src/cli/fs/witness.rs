use std::path::{Path, PathBuf};

use acvm::acir::native_types::WitnessStack;
use nargo::constants::WITNESS_EXT;

use super::{create_named_dir, write_to_file};
use crate::errors::FilesystemError;

pub(crate) fn save_witness_to_dir<P: AsRef<Path>>(
    witness_stack: WitnessStack,
    witness_name: &str,
    witness_dir: P,
) -> Result<PathBuf, FilesystemError> {
    create_named_dir(witness_dir.as_ref(), "witness");
    let witness_path = witness_dir.as_ref().join(witness_name).with_extension(WITNESS_EXT);

    let buf: Vec<u8> = witness_stack.try_into()?;

    write_to_file(buf.as_slice(), &witness_path);

    Ok(witness_path)
}
