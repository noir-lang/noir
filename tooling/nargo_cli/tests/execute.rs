#[allow(unused_imports)]
#[cfg(test)]
mod tests {
    // Some of these imports are consumed by the injected tests
    use assert_cmd::prelude::*;
    use predicates::prelude::*;

    use std::collections::BTreeMap;
    use std::fs;
    use std::path::PathBuf;
    use std::process::Command;

    use super::*;

    test_binary::build_test_binary_once!(mock_backend, "../backend_interface/test-binaries");

    // include tests generated by `build.rs`
    include!(concat!(env!("OUT_DIR"), "/execute.rs"));
}
