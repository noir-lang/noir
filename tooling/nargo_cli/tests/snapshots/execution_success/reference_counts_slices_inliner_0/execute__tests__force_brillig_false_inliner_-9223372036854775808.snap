---
source: tooling/nargo_cli/tests/execute.rs
expression: artifact
---
{
  "noir_version": "[noir_version]",
  "hash": "[hash]",
  "abi": {
    "parameters": [],
    "return_type": null,
    "error_types": {
      "17843811134343075018": {
        "error_kind": "string",
        "string": "Stack too deep"
      }
    }
  },
  "bytecode": "H4sIAAAAAAAA/+2ZS4/aMBDH7SU8El5Z2LZ8gb0HAoTeOCyH9lukS5F6qaqqUi+ttB+9jepJzHS2HPiPNtCMhPKw+c3D9sSZWFNJ5I6BO96Y52Xrjsl5MldkJyXwShxKr82h5bU5tG4cqrlD2ZFD9vId2hw5ZC7foWYN1d2hKm1bz6lTQk6/d8eQ3S9YLWAAQo+L5m+S5Tpk/oHtT0OPqcBfhGz8wPyk6zgPT8d8w/S2WD/+n0KG7tyav/tQm7+odq6t9/s3MtX52J2HjKcxN3yb0LGdCvaTrr7z9YH56sfewmzJPkjzCMdflPO0rcJfl/Z3dPhvid/V4X8kfk+HvyR+qMJPc+JHOvaXa7Gvw18Rf6DDfyT+UIWf7Sl/Up4kHcZUuWPs3cflsfncMn3GHOcpw/RHzFb0nsEyfWQPjw/lWopdLNgaC208B8eCnljQI7ECIKsNZHWArC6Q1QOykLEPa2pXBGQh50QfyBoAWXWdX/TckPaBhWzdMTlTLLPdt1XKnUXsKe98+frp8zfDxAdZz2C6vmFtgRcEv1/rGZ45wbcCq5BiAtKiDZ6O9Wzd/eQMyTZ/gllIy/HbpvLDl8Br9/u/c9f+IqZjcIadhyyfH9L8kK/y/X75mE8YvxD/ReSnO29e9v8pzcu+qcaO+uy8PjvGPacgQBu04pzW9/9QEPjOfG0KApI0BYET/AsvCGQr3YJAlioXBDLdgkBV0BjpxGcpFRWA/DI/xCr8RVmQudWxvxzfic74bog/VeGn5fy/0+GX4/tKJ/5lQey1qYSeq6T7jXcf9xxfrC3TZ4xcECP9EbMVva+2TB/Zw+PDC2IzwdZYaON7kJmgZybokVgBkNUGsjpAVhfI6gFZyNiHNbUrArKQ49gHspBzdQBkIeM1BLKQa6iueWIEZCFjj5xfSLvGQBYyf8U1tesWyEKuIeTaRq6hCZBV1+f2FMi6A7LoneMlP6ZI+261jyn3rN+pjynWM9QKOu+F/13CB5Uf7vqlP6j8AnIYlTqwMQAA",
  "debug_symbols": "7dndaiIxGMbxe8nxHOR98+2tLEuxOi0DouLHwiLe+86Ujbu0g/LX4tEcRvPoU3/BxuRklu3r8f2lW79t9mb242RWm8X80G3W/ehk7MdD++18PYz2h/nuYGY5xca066WZFdFzY966VWtmwZ6bL1PFS/o7V7zz1yfnIvV11V2mqp5/NkbGihTJNRDdjSIh2Vok5HR/ER0tki5/pZVwo0nWUidnH+9v4saaiFW9VMk3PhQVqZIqj+j4b6iS66eiah/wCaNV1PpaReMNIPWlTtZwa4FfqxIfrxIuK1yDe2CtpPEq5QLkbLn+6s6G2tuJ9U9aWPnZvb9pFZbHe0usVZxa+5TvlH7wuutWq+795fN/gZOR0eV89Ts74UTGiUITanFCcEJxwuGEx4mAE9hcsblic8XmDps7bO6wucPmDps7bO6wucPmDps7bO6xucfmHpt7bO6xucfmHpt7bO6xucfmAZsHbB6wecDmAZsHbB6wecDmAZsHbB6xecTmEZtHbB5HzZOvvyxTtp8TASciS5w/fnifpp3ctJObdnLTTm7ayU07uWknN+3kpp3cHTu5qwlsHrF5xOYRm0dsnrB5wuYJmydsnrB5wuYJmydsnrB5wuYZm2dsnrF5xuYZm2dsnrF5xuYZm2dsXrB5weYFmxdsXrB5weYFmxdsXrB5weZiLY8IjyiPOB7xPBJ4hF+yWX42Y/nhjOX6wvWF6wvXF64vXF/w6ZxI5JEET/T60a/5rpu/rtrhbnd48rhe1Kvefnj4va3P1Mvg7W6zaJfHXTtcC/93IzwAx9wk9++gcHhIpDSivn+v/v3+AA==",
  "file_map": {
    "22": {
      "source": "pub mod hash;\npub mod aes128;\npub mod array;\npub mod slice;\npub mod ecdsa_secp256k1;\npub mod ecdsa_secp256r1;\npub mod embedded_curve_ops;\npub mod field;\npub mod collections;\npub mod compat;\npub mod convert;\npub mod option;\npub mod string;\npub mod test;\npub mod cmp;\npub mod ops;\npub mod default;\npub mod prelude;\npub mod runtime;\npub mod meta;\npub mod append;\npub mod mem;\npub mod panic;\npub mod hint;\n\nuse convert::AsPrimitive;\n\n// Oracle calls are required to be wrapped in an unconstrained function\n// Thus, the only argument to the `println` oracle is expected to always be an ident\n#[oracle(print)]\nunconstrained fn print_oracle<T>(with_newline: bool, input: T) {}\n\nunconstrained fn print_unconstrained<T>(with_newline: bool, input: T) {\n    print_oracle(with_newline, input);\n}\n\npub fn println<T>(input: T) {\n    // Safety: a print statement cannot be constrained\n    unsafe {\n        print_unconstrained(true, input);\n    }\n}\n\npub fn print<T>(input: T) {\n    // Safety: a print statement cannot be constrained\n    unsafe {\n        print_unconstrained(false, input);\n    }\n}\n\npub fn verify_proof<let N: u32, let M: u32, let K: u32>(\n    verification_key: [Field; N],\n    proof: [Field; M],\n    public_inputs: [Field; K],\n    key_hash: Field,\n) {\n    verify_proof_internal(verification_key, proof, public_inputs, key_hash, 0);\n}\n\npub fn verify_proof_with_type<let N: u32, let M: u32, let K: u32>(\n    verification_key: [Field; N],\n    proof: [Field; M],\n    public_inputs: [Field; K],\n    key_hash: Field,\n    proof_type: u32,\n) {\n    if !crate::runtime::is_unconstrained() {\n        crate::assert_constant(proof_type);\n    }\n    verify_proof_internal(verification_key, proof, public_inputs, key_hash, proof_type);\n}\n\n#[foreign(recursive_aggregation)]\nfn verify_proof_internal<let N: u32, let M: u32, let K: u32>(\n    verification_key: [Field; N],\n    proof: [Field; M],\n    public_inputs: [Field; K],\n    key_hash: Field,\n    proof_type: u32,\n) {}\n\n// Asserts that the given value is known at compile-time.\n// Useful for debugging for-loop bounds.\n#[builtin(assert_constant)]\npub fn assert_constant<T>(x: T) {}\n\n// Asserts that the given value is both true and known at compile-time\n#[builtin(static_assert)]\npub fn static_assert<let N: u32>(predicate: bool, message: str<N>) {}\n\n#[deprecated(\"wrapping operations should be done with the Wrapping traits. E.g: x.wrapping_add(y)\")]\npub fn wrapping_add<T>(x: T, y: T) -> T\nwhere\n    T: AsPrimitive<Field>,\n    Field: AsPrimitive<T>,\n{\n    AsPrimitive::as_(x.as_() + y.as_())\n}\n#[deprecated(\"wrapping operations should be done with the Wrapping traits. E.g: x.wrapping_sub(y)\")]\npub fn wrapping_sub<T>(x: T, y: T) -> T\nwhere\n    T: AsPrimitive<Field>,\n    Field: AsPrimitive<T>,\n{\n    //340282366920938463463374607431768211456 is 2^128, it is used to avoid underflow\n    AsPrimitive::as_(x.as_() + 340282366920938463463374607431768211456 - y.as_())\n}\n#[deprecated(\"wrapping operations should be done with the Wrapping traits. E.g: x.wrapping_mul(y)\")]\npub fn wrapping_mul<T>(x: T, y: T) -> T\nwhere\n    T: AsPrimitive<Field>,\n    Field: AsPrimitive<T>,\n{\n    AsPrimitive::as_(x.as_() * y.as_())\n}\n\n#[builtin(as_witness)]\npub fn as_witness(x: Field) {}\n\nmod tests {\n    use super::ops::arith::WrappingMul;\n\n    #[test(should_fail_with = \"custom message\")]\n    fn test_static_assert_custom_message() {\n        super::static_assert(1 == 2, \"custom message\");\n    }\n\n    #[test]\n    fn test_wrapping_mul() {\n        let zero: u128 = 0;\n        let one: u128 = 1;\n        let two_pow_64: u128 = 0x10000000000000000;\n        let u128_max: u128 = 0xffffffffffffffffffffffffffffffff;\n\n        // 1*0==0\n        assert_eq(zero, zero.wrapping_mul(one));\n\n        // 0*1==0\n        assert_eq(zero, one.wrapping_mul(zero));\n\n        // 1*1==1\n        assert_eq(one, one.wrapping_mul(one));\n\n        // 0 * ( 1 << 64 ) ==  0\n        assert_eq(zero, zero.wrapping_mul(two_pow_64));\n\n        // ( 1 << 64 ) * 0 == 0\n        assert_eq(zero, two_pow_64.wrapping_mul(zero));\n\n        // 1 * ( 1 << 64 ) == 1 << 64\n        assert_eq(two_pow_64, two_pow_64.wrapping_mul(one));\n\n        // ( 1 << 64 ) * 1 == 1 << 64\n        assert_eq(two_pow_64, one.wrapping_mul(two_pow_64));\n\n        // ( 1 << 64 ) * ( 1 << 64 ) == 1 << 64\n        assert_eq(zero, two_pow_64.wrapping_mul(two_pow_64));\n        // -1 * -1 == 1\n        assert_eq(one, u128_max.wrapping_mul(u128_max));\n    }\n}\n",
      "path": "std/lib.nr"
    },
    "50": {
      "source": "// This test is exactly the same as `reference_counts_inliner_0` which uses\n// arrays rather than slices.\n// This test exists to make sure that our reference counting debug methods match\n// between arrays and slices.\n// We could most likely combine the code for these tests (e.g. using generics),\n// but it is simpler to debug isolated tests.\n// It should only be necessary to have a test at one inliner setting, as we\n// are just checking for discrepancies between the array and slice debugging builtin functions.\n// The actual functionality of reference counting is tested with the `reference_counts_*` tests.\n// We went with testing at an inliner aggressiveness of zero, as this is generally\n// the most useful inliner setting for unconstrained functions.\nuse std::mem::slice_refcount;\n\nfn main() {\n    let mut slice = &[0, 1, 2];\n    assert_refcount(slice, 1, true);\n\n    borrow(slice, slice_refcount(slice));\n    borrow_mut(&mut slice, slice_refcount(slice));\n    let _ = copy_mut(slice, slice_refcount(slice));\n\n    borrow_mut_two(&mut slice, &mut slice, slice_refcount(slice));\n\n    let mut u32_slice = &[0, 1, 2];\n    let rc1 = slice_refcount(slice);\n    let rc2 = slice_refcount(u32_slice);\n    borrow_mut_two_separate(&mut slice, &mut u32_slice, rc1, rc2);\n\n    // Safety: test\n    regression_7297();\n}\n\nfn borrow(slice: [Field], rc_before_call: u32) {\n    assert_refcount(slice, rc_before_call, true);\n    println(slice[0]);\n}\n\nfn borrow_mut(slice: &mut [Field], rc_before_call: u32) {\n    assert_refcount(*slice, rc_before_call, true);\n    slice[0] = 3;\n    println(slice[0]);\n}\n\n// Returns a new slice (a copy) to prevent SSA from optimizing away mutations.\nfn copy_mut(mut slice: [Field], rc_before_call: u32) -> [Field] {\n    assert_refcount(slice, rc_before_call, true);\n    slice = &[4, slice[1], slice[2]];\n    println(slice[0]);\n    slice\n}\n\nfn borrow_mut_two(slice1: &mut [Field], slice2: &mut [Field], rc_before_call: u32) {\n    assert_refcount(*slice1, rc_before_call, true);\n    assert_refcount(*slice2, rc_before_call + 1, true); // should be a copy\n    slice1[0] = 5;\n    slice2[0] = 6;\n    println(slice1[0]); // slice1 & 2 alias, so this should also print 6\n    println(slice2[0]);\n}\n\nfn borrow_mut_two_separate(\n    slice1: &mut [Field],\n    slice2: &mut [u32],\n    rc_before_call1: u32,\n    rc_before_call2: u32,\n) {\n    assert_refcount(*slice1, rc_before_call1, true);\n    assert_refcount(*slice2, rc_before_call2, true);\n    slice1[0] = 7;\n    slice2[0] = 8;\n    println(slice1[0]);\n    println(slice2[0]);\n}\n\nfn assert_refcount<T>(slice: [T], mut expected: u32, expect_copy: bool) {\n    let count = slice_refcount(slice);\n\n    if expect_copy {\n        expected += 1;\n    }\n\n    if std::runtime::is_unconstrained() {\n        if count != expected {\n            println(f\"actual = {count}, expected = {expected}\");\n        }\n        assert_eq(count, expected);\n    } else {\n        assert_eq(count, 0);\n    }\n}\n\nfn regression_7297() {\n    let mut slice: [Field] = &[0, 1, 2];\n\n    let refcount_0 = slice_refcount(slice);\n    borrow_mut_two(&mut slice, &mut slice, refcount_0);\n\n    let refcount_1 = slice_refcount(slice);\n    let slice_2 = copy_mut(slice, refcount_1 + 1);\n    let refcount_2 = slice_refcount(slice);\n\n    assert_eq(slice[0], 6, \"the original should not be mutated by copy_mut, only borrow_mut_two\");\n    assert_eq(slice_2[0], 4, \"the copy should have the expected content\");\n\n    if std::runtime::is_unconstrained() {\n        assert(\n            refcount_1 != 0,\n            \"borrow_mut_two should create a fresh slice and not decrease its RC\",\n        );\n\n        assert_eq(\n            refcount_1,\n            2,\n            \"There is 1 clone after `borrow_mut_two` and before `refcount_1` is defined (cloned before slice_refcount call)\",\n        );\n        assert_eq(\n            refcount_2,\n            refcount_1 + 3,\n            \"after refcount_1 we clone once in passing slice to copy_mut, once to slice_refcount after, and once within copy_mut\",\n        );\n    }\n}\n",
      "path": ""
    }
  },
  "names": [
    "main"
  ],
  "brillig_names": [
    "print_unconstrained",
    "print_unconstrained"
  ]
}
